### PR TITLE
ceph-dencoder: Add global to denc-mod-osd's target_link_libraries

### PR DESCRIPTION
**Pull request by tserong**
Fixes: https://tracker.ceph.com/issues/57390
Signed-off-by: Tim Serong <tserong@suse.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by tserong** [comment]
Note: pacific does't need this one. It's only necessary on main and quincy.

**Comment by tserong** [comment]
Ping @tchaikov :-)

File: src/tools/ceph-dencoder/CMakeLists.txt [hunk 1]
```diff
@@ -50,7 +50,8 @@ target_link_libraries(denc-mod-osd
   os
   osd
   mon
-  erasure_code)
+  erasure_code
+  global)
```
Comment (tchaikov) [inline src/tools/ceph-dencoder/CMakeLists.txt:54]:
```suggestion
  erasure_code
  global-static)
```

see 4ea38df656b2aad9398fcb1656c294f5b8366811 for the rationales.

Comment (tserong) [reply src/tools/ceph-dencoder/CMakeLists.txt:54]: I've just tried that, and now `ceph-dencoder` segfaults:

```
# ceph-dencoder 
-h for help
Segmentation fault (core dumped)

# gdb ceph-dencoder 
[...]
Reading symbols from ceph-dencoder...
(No debugging symbols found in ceph-dencoder)
(gdb) run
Starting program: /usr/bin/ceph-dencoder 
Missing separate debuginfos, use: zypper install ceph-common-debuginfo-17.0.0-14884.1.x86_64
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
-h for help

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff7de1123 in tcmalloc::ThreadCache::ReleaseToCentralCache(tcmalloc::ThreadCache::FreeList*, unsigned int, int) () from /lib64/libtcmalloc.so.4

```

Note that ceph-dencoder itself links against `global`, not `global-static` (see https://github.com/ceph/ceph/blob/a1bf7f7416f643f8b56fb571f687fa90066cbd86/src/tools/ceph-dencoder/CMakeLists.txt#L99). Could that discrepancy cause this? Should I try changing that to global-static as well, or go back to using `global` for denc-mod-osd?

Comment (tserong) [reply src/tools/ceph-dencoder/CMakeLists.txt:54]: Just for a test, I tried using `global-static` on denc-mod-osd as above, and switched `global` to `global-static` in ceph-dencoder like so:

```diff
--- ceph-17.0.0-14883-gc49b81c7d61.orig/src/tools/ceph-dencoder/CMakeLists.txt
+++ ceph-17.0.0-14883-gc49b81c7d61/src/tools/ceph-dencoder/CMakeLists.txt
@@ -97,7 +97,7 @@ target_compile_definitions(ceph-dencoder
 
 target_link_libraries(ceph-dencoder
   StdFilesystem::filesystem
-  global
+  global-static
   ${DENCODER_EXTRALIBS}
   cls_log_client
   cls_version_client
```

That fixes the segfault, but now I've got another undefined symbol (this time in denc-mod-common):

```
# ceph-dencoder 
failed to dlopen("/usr/lib64/ceph/denc/denc-mod-common.so"): /usr/lib64/ceph/denc/denc-mod-common.so: undefined symbol: _ZTI7Message
-h for help
```

Based on these tests, I think we just need to use `global`, not `global-static`, unless there's some other change(s) necessary in addition to the above.

Comment (tchaikov) [reply src/tools/ceph-dencoder/CMakeLists.txt:54]: @tserong hi Tim, sorry. i didn't look into the linkage of ceph-dencoder itself. agreed, it it links against `global-static`, we should not link `denc-mod-*` against it again. it'd violate the ODR (one definition rule). but after a second thought, i think we need to drop `global` from ceph-dencoder's linked libraries list. BTW, sorry for the latency. i was in a "power saver mode" last month.

Comment (tchaikov) [reply src/tools/ceph-dencoder/CMakeLists.txt:54]: #48418 is created to implement this proposal.

**Review by tchaikov — APPROVED** [review]
@tserong the `global` linkage is likely to be polled in by `mon` and `osd` which reference `ErasureCodePluginRegistry::instance()`. but i still don't understand why this makes it a dependency. as the function using this symbol is not used by `ceph-dencoder` or any of its `denc-mod-*` plugins. guess LTO thinks different than I do..

but i don't want to block you because of my incompetence. from the test result and your reasoning, this change makes sense.

**Comment by tserong** [comment]
> BTW, sorry for the latency. i was in a "power saver mode" last month.

No problem!

> but i don't want to block you because of my incompetence. from the test result and your reasoning, this change makes sense.

Incompetence? Nah man, you still know more about this stuff than me :-)

**Pull request merged**